### PR TITLE
Plan stock rebuild

### PR DESCRIPTION
**Pull request by dannyroberts**
This adds columns to the stock state view that shows projected changes from rebuilding state for a particular stock.

(This view is still totally not scalable, as it just loads every stock transaction of a domain at once. For the purposes of the domain I'm looking at, though, it's fine, and we can figure out what we want to do in terms of pagination later.)

It does this by (as you can see in the first two commits) breaking the rebuild function into two steps—planning and executing—and then only calling the planning part and reporting the proposed modifications.

### CODE REVIEW

**Comment by dannyroberts** [comment]
@czue @TylerSheffels FYI

File: corehq/apps/commtrack/processing.py [hunk 1]
```diff
@@ -201,27 +202,35 @@ def _adjust_ledger_values(original_balance, stock_transaction):
...
-    and the quantity and stock_on_hand fields of StockTransaction
-    when they are calculated from previous state
-    (as opposed to part of the explict transaction)
+    planner for rebuild_stock_state
 
+    yields actions for rebuild_stock_state to take,
+    facilitating doing a dry run
+
+    Warning: since some important things are still done through signals
+    rather than here explicitly,
```
Comment (czue) [inline corehq/apps/commtrack/processing.py]: i feel like the second half of this thought might be important...

File: corehq/apps/commtrack/templates/commtrack/manage/rebuild_stock_state.html [hunk 2]
```diff
@@ -3,39 +3,78 @@
...
+                    <input type="hidden" name="product_id" value="{{ key.product_id }}">
+                    <input type="submit" class="btn btn-primary" value="Rebuild"/>
+                </form>
+                {% endif %}
+            </td>
+        </tr>
+        <tr>
+            <th colspan="2"></th>
+            <th colspan="2">Current Values</th>
+            <th colspan="2">Rebuilt Values</th>
```
Comment (czue) [inline corehq/apps/commtrack/templates/commtrack/manage/rebuild_stock_state.html]: i know we expect this page to be internal, but it doesn't hurt to throw `trans` tags on these

File: corehq/apps/commtrack/views.py [hunk 3]
```diff
@@ -293,12 +296,35 @@ class RebuildStockStateView(BaseCommTrackManageView):
 
     @property
     def page_context(self, **kwargs):
```
Comment (czue) [inline corehq/apps/commtrack/views.py]: should we consider limiting what this returns explicitly? just worried about someone trying to load this page on some big domain and then the server running out of memory.

**Comment by czue** [comment]
no blockers but will give you a chance to respond before merging

File: corehq/apps/commtrack/processing.py (hunk 1 above)
Comment (dannyroberts) [inline corehq/apps/commtrack/processing.py]: Nah, I'll just leave it up to your imagination...

File: corehq/apps/commtrack/templates/commtrack/manage/rebuild_stock_state.html (hunk 2 above)
Comment (dannyroberts) [inline corehq/apps/commtrack/templates/commtrack/manage/rebuild_stock_state.html]: Fair enough

File: corehq/apps/commtrack/views.py (hunk 3 above)
Comment (dannyroberts) [inline corehq/apps/commtrack/views.py]: Yeah that's a good idea. Done

**Comment by dannyroberts** [comment]
Ok this is all set on my end

**Comment by TylerSheffels** [comment]
Looks good to me, and +1 for the change

**Comment by dannyroberts** [comment]
@czue I think this is good to merge as is, and I can do further changes on top of it. (Since it's all transfers, and transfers are commutative, running this for moko as is will put them at the right total balance, even if the individual transactions are out of order.)

**Pull request merged**